### PR TITLE
core/types: remove Message

### PR DESCRIPTION
**Pull request by roberto-bayardo**
This is a simple cleanup PR, the goal of which is to encapsulate all the Message related structures & abstractions within a single package (core). Previously it was leaking into core/types/Transaction.go, most of whose users have no need for it.

I've also changed Message to be a simple struct rather than a struct + an interface, as the interface was only adding unnecessary complexity.

### CODE REVIEW

**Comment by fjl** [comment]
I like this, but would prefer not to have the `NewMessage` constructor. We will for sure have to add/change parameters of this function in the future. If they are passed by position, it can create a lot of extra changes.

**Comment by roberto-bayardo** [comment]
@fjl here's how the followup PR which removes NewMessage and exports the struct members would look: https://github.com/roberto-bayardo/go-ethereum/pull/1

File: internal/ethapi/transaction_args.go [hunk 1]
```diff
@@ -199,10 +200,10 @@ func (args *TransactionArgs) setLondonFeeDefaults(ctx context.Context, head *typ
 // ToMessage converts the transaction arguments to the Message type used by the
 // core evm. This method is used in calls and traces that do not require a real
 // live transaction.
-func (args *TransactionArgs) ToMessage(globalGasCap uint64, baseFee *big.Int) (types.Message, error) {
+func (args *TransactionArgs) ToMessage(globalGasCap uint64, baseFee *big.Int) (*core.Message, error) {
 	// Reject invalid combinations of pre- and post-1559 fee styles
 	if args.GasPrice != nil && (args.MaxFeePerGas != nil || args.MaxPriorityFeePerGas != nil) {
-		return types.Message{}, errors.New("both gasPrice and (maxFeePerGas or maxPriorityFeePerGas) specified")
+		return &core.Message{}, errors.New("both gasPrice and (maxFeePerGas or maxPriorityFeePerGas) specified")
```
Comment (rjl493456442) [inline internal/ethapi/transaction_args.go:206]: return nil, errors.New("xxxx")

Comment (roberto-bayardo) [reply internal/ethapi/transaction_args.go]: changed

**Review by rjl493456442 — APPROVED** [review]
Thanks, looks pretty good to me!

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM

**Comment by fjl** [comment]
Sorry, I will get to re-reviewing this PR soon!

**Comment by roberto-bayardo** [comment]
@fjl wdyt?

**Comment by fjl** [comment]
I looked into your draft PR to remove the accessor methods, and we should get those changes in here as well.

**Comment by roberto-bayardo** [comment]
> I looked into your draft PR to remove the accessor methods, and we should get those changes in here as well.

cool, will merge them in.

**Comment by roberto-bayardo** [comment]
> I looked into your draft PR to remove the accessor methods, and we should get those changes in here as well.

done

**Comment by fjl** [comment]
I've rebased this, and have done some more renames.

**Comment by roberto-bayardo** [comment]
> I've rebased this, and have done some more renames.

lgtm

**Comment by fjl** [comment]
It's in! I added a bit of context to the commit message: 

Here, the core.Message interface turns into a plain struct.

> This is a breaking change to packages core and core/types. While we do
> not promise API stability for package core, we do for core/types. An
> exception can be made for types.Message, since it doesn't have any
> purpose apart from invoking the state transition in package core.
> types.Message was also marked deprecated in the same commit it got
> added, 4dca5d4db7 (November 2016).
>
> The core.Message interface was added in December 2014, in commit
> db494170dc, for the purpose of 'testing' state transitions. It's the
> same commit that made transaction struct fields private. Before that
> change, the state transition was always invoked on full transactions.
> 
> Over time, multiple implementations of the interface accrued across
> different packages, since constructing a Message is required whenever
> one wants to invoke the state transition. These implementations all
> looked very similar, a struct with private fields exposing the fields
> as accessor methods.
>
> By changing Message into a struct with public fields we can remove all
> these useless interface implementations. It will also hopefully
> simplify future changes to the type with less updates to apply across
> all of go-ethereum when a field is added to Message.

**Pull request merged**